src/iarc7_motion/iarc_tasks/task_utilities/pid_controller.py

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python
+import rospy
+import numpy as np
+
+class PidSettings(object): 
+    def __init__(self, settings_dict):
+        self.kp = settings_dict.get('kp', 0)
+        self.ki = settings_dict.get('ki', 0)
+        self.kd = settings_dict.get('kd', 0)
+        self.accum_max = settings_dict.get('accumulator_max', 0)
+        self.accum_min = settings_dict.get('accumulator_min', 0)
+        self.accum_en_threshold = settings_dict.get('accumulator_enable_threshold', 0)
+
+class PidController(object):
+    def __init__(self, settings):
+        self._p_gain = settings.kp
+        self._i_gain = settings.ki
+        self._d_gain = settings.kd
+        self._i_accumulator_max = settings.accum_max
+        self._i_accumulator_min = settings.accum_min
+        self._i_accumulator_enable_threshold = settings.accum_en_threshold
+        
+        self._initialized = False
+        self._last_time = None
+
+        self._i_accumulator = 0.0
+        self._last_current_value = 0.0
+        self._setpoint = 0.0
+
+    def reset_accumulator(self):
+        self._i_accumulator = 0.0
+
+    def set_setpoint(self, setpoint):
+        self._setpoint = setpoint
+
+    def update(self, current_value, time, log_debug):
+        '''
+        Updates PID controller
+
+        Args:
+            current_value: current value of whatever you are controlling on
+            time: current time; use rospy.get_time()
+            log_debug: enables verbose debugging
+        
+        Returns: 
+            success: whether or not a response could be calculated
+            response is the filter output
+        '''
+        response = 0
+
+        if self._last_time is None:
+            self._last_time = time
+            self._initialized = True
+            return True, response
+        
+        if time == self._last_time:
+            rospy.logwarn(
+                    'Time passed in to PidController is equal to last time.')
+            return False, response
+
+        if time < self._last_time:
+            rospy.logwarn(
+                    'Time passed in to PidController is less than the last time.')
+            return False, response
+
+        if not np.isfinite(current_value):
+            rospy.logwarn(
+                    'Invalid argument to PidControllor.update (current_value = {}).'.format(current_value))
+            return False, response
+
+        difference = self._setpoint - current_value
+        p_term = self._p_gain * difference
+        response = p_term
+
+        if self._initialized:
```

Comment:
Why do we need to check this here? This should always be true if we get to this point

Reply:
Fixed